plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -144,17 +166,71 @@ codeActionResolveProvider ideState pId ca uri uid = do
   -- that this resolve is stale.
   record <- handleMaybe PluginStaleResolve $ IntMap.lookup uid crCodeActionResolve
   -- We should never fail to render
-  rendered <- handleMaybe (PluginInternalError "Failed to render") $ renderRecordInfo nameMap record
+  rendered <- handleMaybe (PluginInternalError "Failed to render") $ renderRecordInfoAsTextEdit nameMap record
   let edits = [rendered]
               <> maybeToList (pragmaEdit enabledExtensions pragma)
   pure $ ca & L.edit ?~ mkWorkspaceEdit edits
   where
     mkWorkspaceEdit ::[TextEdit] -> WorkspaceEdit
     mkWorkspaceEdit edits = WorkspaceEdit (Just $ Map.singleton uri edits) Nothing Nothing
-    pragmaEdit :: [Extension] -> NextPragmaInfo -> Maybe TextEdit
-    pragmaEdit exts pragma = if NamedFieldPuns `elem` exts
-                      then Nothing
-                      else Just $ insertNewPragma pragma NamedFieldPuns
+
+inlayHintProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentInlayHint
+inlayHintProvider _ state pId InlayHintParams {_textDocument = TextDocumentIdentifier uri, _range = visibleRange} = do
+  nfp <- getNormalizedFilePathE uri
+  pragma <- getFirstPragma pId state nfp
+  runIdeActionE "ExplicitFields.CollectRecords" (shakeExtras state) $ do
+    (crr@CRR {crCodeActions, crCodeActionResolve}, pm) <- useWithStaleFastE CollectRecords nfp
+    let -- Get all records with dotdot in current nfp
+        records = [ record
+                  | Just range <- [toCurrentRange pm visibleRange]
+                  , uid <- RangeMap.flippedFilterByRange range crCodeActions
+                  , Just record <- [IntMap.lookup uid crCodeActionResolve] ]
+        -- Get the definition of each dotdot of record
+        locations = [ getDefinition nfp pos
+                    | record <- records
+                    , pos <- maybeToList $ fmap _start $ recordInfoToDotDotRange record ]
+    defnLocsList <- liftIO $ Shake.runIdeAction "ExplicitFields.getDefinition" (shakeExtras state) (sequence locations)
+    pure $ InL $ mapMaybe (mkInlayHints crr pragma) (zip defnLocsList records)
+   where
+     mkInlayHints :: CollectRecordsResult -> NextPragmaInfo -> (Maybe [(Location, Identifier)], RecordInfo) -> Maybe InlayHint
+     mkInlayHints CRR {enabledExtensions, nameMap} pragma (defnLocs, record) =
+       let range = recordInfoToDotDotRange record
+           textEdits = maybeToList (renderRecordInfoAsTextEdit nameMap record)
+                    <> maybeToList (pragmaEdit enabledExtensions pragma)
+           names = renderRecordInfoAsLabelName record
+       in do
+         end <- fmap _end range
+         names' <- names
+         defnLocs' <- defnLocs
+         let excludeDotDot (Location _ (Range _ pos)) = pos /= end
+             -- find location from dotdot definitions that name equal to label name
+             findLocation t = fmap fst . find (either (const False) ((==) t) . snd) . filter (excludeDotDot . fst)
```

Comment:
This feels hard to read, could you add explicit parameters and intermediate results?